File: v1/HLD/Caching/code/LruCache.cpp
#include <iostream>
#include <thread>
#include <chrono>
#include <list>
#include <unordered_map>
#include <vector>
#include <string>

const int SOURCE_LATENCY_MS = 10;
const int CAPACITY = 3;
const std::vector<int> KEY_SEQUENCE = {1, 2, 3, 1, 2, 4, 1, 5, 1, 2};

class LRUCache {
public:
    LRUCache(int capacity) : capacity(capacity) {}

    bool get(int key, std::string& value) {
        auto it = index.find(key);
        if (it == index.end()) return false;
        order.splice(order.begin(), order, it->second);
        value = it->second->second;
        return true;
    }

    void put(int key, const std::string& value) {
        auto it = index.find(key);
        if (it != index.end()) {
            order.erase(it->second);
        }
        order.push_front({key, value});
        index[key] = order.begin();
        if ((int)order.size() > capacity) {
            auto last = order.back();
            std::cout << "  evicted key " << last.first << " (least recently used)" << std::endl;
            index.erase(last.first);
            order.pop_back();
        }
    }

private:
    int capacity;
    std::list<std::pair<int, std::string>> order;
    std::unordered_map<int, std::list<std::pair<int, std::string>>::iterator> index;
};

std::string slowFetch(int key) {
    std::this_thread::sleep_for(std::chrono::milliseconds(SOURCE_LATENCY_MS));
    return "value-" + std::to_string(key);
}

double runWithCache(int& hits, int& misses) {
    LRUCache cache(CAPACITY);
    hits = 0;
    misses = 0;
    auto start = std::chrono::steady_clock::now();
    for (int key : KEY_SEQUENCE) {
        std::string value;
        if (cache.get(key, value)) {
            hits++;
        } else {
            misses++;
            value = slowFetch(key);
            cache.put(key, value);
        }
    }
    return std::chrono::duration<double, std::milli>(std::chrono::steady_clock::now() - start).count();
}

double runWithoutCache() {
    auto start = std::chrono::steady_clock::now();
    for (int key : KEY_SEQUENCE) {
        slowFetch(key);
    }
    return std::chrono::duration<double, std::milli>(std::chrono::steady_clock::now() - start).count();
}

int main() {
    std::cout << "With LRU cache:" << std::endl;
    int hits, misses;
    double withCacheMs = runWithCache(hits, misses);
    std::cout << "  hits=" << hits << " misses=" << misses << " total_time=" << withCacheMs << " ms" << std::endl;

    double withoutCacheMs = runWithoutCache();
    std::cout << "Without cache: total_time=" << withoutCacheMs << " ms" << std::endl;

    std::cout << "Speedup from caching: " << (withoutCacheMs / withCacheMs) << "x" << std::endl;
    return 0;
}

File: v1/HLD/Caching/code/LruCache.java
import java.util.LinkedHashMap;
import java.util.Map;

public class LruCache {
    static final int SOURCE_LATENCY_MS = 10;
    static final int CAPACITY = 3;
    static final int[] KEY_SEQUENCE = {1, 2, 3, 1, 2, 4, 1, 5, 1, 2};

    static class Cache extends LinkedHashMap<Integer, String> {
        int capacity;
        Cache(int capacity) {
            super(16, 0.75f, true);
            this.capacity = capacity;
        }
        @Override
        protected boolean removeEldestEntry(Map.Entry<Integer, String> eldest) {
            boolean evict = size() > capacity;
            if (evict) {
                System.out.println("  evicted key " + eldest.getKey() + " (least recently used)");
            }
            return evict;
        }
    }

    static String slowFetch(int key) throws InterruptedException {
        Thread.sleep(SOURCE_LATENCY_MS);
        return "value-" + key;
    }

    static double runWithCache(int[] hitsMisses) throws InterruptedException {
        Cache cache = new Cache(CAPACITY);
        int hits = 0, misses = 0;
        long start = System.nanoTime();
        for (int key : KEY_SEQUENCE) {
            String value = cache.get(key);
            if (value != null) {
                hits++;
            } else {
                misses++;
                value = slowFetch(key);
                cache.put(key, value);
            }
        }
        hitsMisses[0] = hits;
        hitsMisses[1] = misses;
        return (System.nanoTime() - start) / 1_000_000.0;
    }

    static double runWithoutCache() throws InterruptedException {
        long start = System.nanoTime();
        for (int key : KEY_SEQUENCE) {
            slowFetch(key);
        }
        return (System.nanoTime() - start) / 1_000_000.0;
    }

    public static void main(String[] args) throws Exception {
        System.out.println("With LRU cache:");
        int[] hitsMisses = new int[2];
        double withCacheMs = runWithCache(hitsMisses);
        System.out.printf("  hits=%d misses=%d total_time=%.2f ms%n", hitsMisses[0], hitsMisses[1], withCacheMs);

        double withoutCacheMs = runWithoutCache();
        System.out.printf("Without cache: total_time=%.2f ms%n", withoutCacheMs);

        System.out.printf("Speedup from caching: %.2fx%n", withoutCacheMs / withCacheMs);
    }
}
